src/app/clusters/color-control-server/color-control-server.cpp

```diff
@@ -1498,12 +1594,135 @@ bool emberAfColorControlClusterStopMoveStepCallback(EndpointId aEndpoint, app::C
-                                                    uint8_t optionsMask, uint8_t optionsOverride)
+static void startColorLoop(EndpointId endpoint, uint8_t startFromStartHue)
+{
+    uint8_t direction   = readColorLoopDirectionAttribute(endpoint);
+    uint16_t time       = readColorLoopTimeAttribute(endpoint);
+    uint16_t currentHue = readEnhancedHue(endpoint);
+    u_int16_t startHue  = startFromStartHue ? readColorLoopStartEnhancedHueAttribute(endpoint) : currentHue;
+
+    writeColorLoopStoredEnhancedHueAttribute(endpoint, currentHue);
+    writeColorLoopActiveAttribute(endpoint, COLOR_LOOP_ACTIVE);
+
+    initHueSat(endpoint);
+
+    colorHueTransitionState.isEnhancedHue = true;
+
+    colorHueTransitionState.initialEnhancedHue = startHue;
+    colorHueTransitionState.currentEnhancedHue = currentHue;
+
+    if (direction == COLOR_LOOP_DIRECTION_INCREMENT)
+    {
+        colorHueTransitionState.finalEnhancedHue = static_cast<uint16_t>(startHue - 1);
+    }
+    else
+    {
+        colorHueTransitionState.finalEnhancedHue = static_cast<uint16_t>(startHue + 1);
+    }
+
+    colorHueTransitionState.up     = direction;
+    colorHueTransitionState.repeat = true;
+
+    colorHueTransitionState.stepsRemaining = static_cast<uint16_t>(time * TRANSITION_TIME_1S);
+    colorHueTransitionState.stepsTotal     = static_cast<uint16_t>(time * TRANSITION_TIME_1S);
+    colorHueTransitionState.endpoint       = endpoint;
+
+    writeRemainingTime(endpoint, MAX_INT16U_VALUE);
+    emberEventControlSetDelayMS(&COLOR_HSV_CONTROL, UPDATE_TIME_MS);
+}
+
+bool emberAfColorControlClusterColorLoopSetCallback(chip::EndpointId aEndpoint, chip::app::CommandHandler * commandObj,
+                                                    uint8_t updateFlags, uint8_t action, uint8_t direction, uint16_t time,
+                                                    uint16_t startHue, uint8_t optionsMask, uint8_t optionsOverride)
 {
-    // TODO IMPLEMENT LOGIC
-    emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_UNSUP_COMMAND);
+    EndpointId endpoint = emberAfCurrentEndpoint();
+    uint8_t isColorLoopActive;
+    uint8_t deactiveColorLoop;
+
+    if (!shouldExecuteIfOff(endpoint, optionsMask, optionsOverride))
+    {
+        emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_SUCCESS);
+        return true;
+    }
+
+    isColorLoopActive = readColorLoopActiveAttribute(endpoint);
+    deactiveColorLoop =
+        (updateFlags & EMBER_AF_COLOR_LOOP_UPDATE_FLAGS_UPDATE_ACTION) && (action == EMBER_ZCL_COLOR_LOOP_ACTION_DEACTIVATE);
+
+    if (updateFlags & EMBER_AF_COLOR_LOOP_UPDATE_FLAGS_UPDATE_DIRECTION)
+    {
+        writeColorLoopDirectionAttribute(endpoint, direction);
+
+        // Checks if color loop is active and stays active
+        if (isColorLoopActive && !deactiveColorLoop)
+        {
+            colorHueTransitionState.up                 = direction;
```

Comment:
You set the direction here and then again in the if/else. Keep this one line only